### DISCUSSION TITLE
Merge Global Data Files with addGlobalData?

### DISCUSSION THREAD

**Discussion by devallibus** [11ty/eleventy#1826] [Q&A] [answered]
Hello,

I've been looking for a way to explicitly merge / concatenate the datasets / arrays returned by several Global Data files that are making API calls, without making each file write a file and then merge these files.

addGlobalData seems to maybe be the ticket but the Official Documentation is just not enough to make it clear to me and I haven't found any help on google.

Am I right to suspect you should be able to do it with addGlobalData?

**Comment by pdehaan** [comment, marked as answer]
I haven't tried global data files yet, but note from [the docs](https://www.11ty.dev/docs/data-global-custom/) that it is `Coming soon in v1.0.0` (and the current stable release from npm is still v0.12.1). So information and details from Google is probably pretty rare currently, and it'd only be available if you install Eleventy directly from GitHub, or a canary release from npm.

I imagine you'd be able to merge some stuff from global data files by requiring the files in your .eleventy.js config and then use `.addGlobalData()`, but I'm not sure you'd have access to stuff like collections.

You could probably create a [global data file](https://www.11ty.dev/docs/data-global/) in the current v0.12 that makes multiple API calls in a single file and return a value without needing to write temp files to disk.

Or, since you said you already have several global data files, I have this terrible code:

```js
// ./src/_data/one.js
module.exports = async () => Promise.resolve("One");
```

And 

```js
// ./src/_data/two.js
module.exports = async () => Promise.resolve("Two");
```

Then I can create a third data file which loads the first two async data files, waits for them to finish doing their async things (API fetching, or whatever), then concatenate the results and return a value:

```js
// ./src/_data/three.js
const one = require("./one");
const two = require("./two");

module.exports = async () => {
  // Since each of the other global data files were returning async functions
  // (assuming you're using something like `axios` or `fetch` to get
  // API results from a remote server, we need to use `async..await` 
  // here to wait for the results to finish.
  // This might need to be tweaked to make them fetch simultaneously vs sequentially.
  const $one = await one();
  const $two = await two();
  // Concatenate the two strings from our previous fake-API calls.
  return $one + $two;
};
```

Now I can use the value of the global `three.js` data file in my templates:

```liquid
<p>{{ three }}</p>
<!-- output: "<p>OneTwo</p> -->
```

**Reply by pdehaan** [reply]
I came up with an even worse version too!
Assuming my ./src/_data/one.js and ./src/_data/two.js files return some more complex Arrays:

```js
module.exports = async () => Promise.resolve(["One", 1, "Uno"]);
```

My ./src/_data/three.js can merge the arrays and do some needless processing+sorting:

```js
const one = require("./one");
const two = require("./two");

module.exports = async () => {
  // Use `Promise.all()` to execute in parallel. `res` will be an Array here.
  const res = await Promise.all([
    one(),
    two()
  ]);

  // Merge the arrays from our various API calls via other global data files.
  return []
    // Use the spread syntax (...) to convert the Array values into properties[1] so it merges everything into a single array.
    .concat(...res)
    // Loop over each item and do pointless data processing.
    .map((value) => {
      if (isNaN(value)) {
        // Make the non-numeric values uppercase for zero reason.
        value = value.toString().toUpperCase();
      }
      return value;
    })
    // Sort the merged Array for no reason apart from "because we can".
    .sort();
};
```

[1] https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/Spread_syntax

**Reply by devallibus** [reply]
That is not terrible at all!

Thank you for your help, this is the second time you save my ass. I didn't know you could make it available with the module.exports !

This all comes from the same project as my question before, which by the way I managed to finish with the implementation you helped me with, successfully (only to learn I had to do away with the pagination and my wonderful EleventyComputed shenanigans bc of SEO shenanigans).

I'll circle back to report on both of these once I get it going!

Thank you again @pdehaan !
